orttraining/orttraining/core/graph/adasum_optimizer_graph_builder.cc

```diff
@@ -115,25 +158,37 @@ Status AdasumOptimizerGraphBuilder::BuildInternal(
     return graph.GenerateNodeArgName(base_name);
   };
 
-  const int64_t horovod_reduce_op = opt_graph_config_.horovod_reduce_op;
-
   // add gradient scaling
   ArgDef fused_gradient_argdef;
   const auto total_num_accumulations = opt_graph_config_.gradient_accumulation_steps;
   ORT_RETURN_IF_NOT(total_num_accumulations > 0);
-  const float scale = 1.0f / total_num_accumulations;
-  // No fusion with Adasum
-  const bool fuse_scaling_outputs = false;
+
+  float scale_divisor = total_num_accumulations;
+  //If Adasum GPU hierarchical reduce is used, then divide gradients by local size.
+  if (opt_graph_config_.adasum_reduction_type == AdasumReductionType::GpuHierarchical) {
+    scale_divisor *= opt_graph_config_.local_size;
+  }
+
+  const float scale = 1.0f / scale_divisor;
+  // Only fuse if using hierarchical reduce.
+  const bool fuse_scaling_outputs = opt_graph_config_.adasum_reduction_type == AdasumReductionType::GpuHierarchical ? true: false;
   ORT_RETURN_IF_ERROR(AddGradientScalingNodes(nodearg_name_generator, scale, gradient_argdefs, fused_gradient_argdef, graph_defs,
                                               opt_graph_config_.AllReduceDataType(), fuse_scaling_outputs));
 
+  // add Allreduce for gradients
+  ArgDef reduced_fused_gradient_argdef;
+  if (opt_graph_config_.adasum_reduction_type == AdasumReductionType::GpuHierarchical) {
+   ORT_RETURN_IF_ERROR(AddNcclAllReduceForGradientsWithGroups(gradient_argdefs, fused_gradient_argdef, graph_defs,
+                                                              reduced_fused_gradient_argdef, (int64_t)2/*node local*/));
```

Comment:
As above, using a class enum and the names will help pick up any changes we make in the future (e.g., adding a new group causing renumbering)